src/main/java/org/wildfly/naming/client/remote/RemoteClientTransport.java

```diff
@@ -136,28 +141,38 @@ public void handleMessage(final Channel channel, final MessageInputStream messag
                         return;
                     }
                     int length = mis.readUnsignedByte();
-                    boolean hasOne = false, hasTwo = false;
+                    boolean hasOne = false, hasTwo = false, hasThree = false;
                     for (int i = 0; i < length; i ++) {
                         // Servers present versions >= 2 with the MSB set so that old clients don't get confused.
                         // We strip the MSB to compensate.
-                        int v = mis.readUnsignedByte() & 0x7f;
+                        int v = min(LATEST_VERSION, mis.readUnsignedByte() & 0x7f);
                         if (v == 1) {
                             hasOne = true;
                         } else if (v == 2) {
                             hasTwo = true;
+                        } else if (v == 3) {
+                            hasThree = true;
                         }
                     }
                     int version;
-                    if (hasTwo) {
-                        version = 2;
+                    if (hasThree) {
+                        version = JAKARTAEE_PROTOCOL_VERSION;
+                    } else if (hasTwo) {
+                        version = JAVAEE_PROTOCOL_VERSION;
                     } else if (hasOne) {
                         version = 1;
                     } else {
                         futureResult.setException(new IOException(Messages.log.noCompatibleVersions()));
                         return;
                     }
                     final MarshallingConfiguration configuration = new MarshallingConfiguration();
-                    configuration.setVersion(version == 2 ? 4 : 2);
+                    configuration.setVersion(version >= 2 ? 4 : 2);
+                    if (version < JAKARTAEE_PROTOCOL_VERSION && LATEST_VERSION >= JAKARTAEE_PROTOCOL_VERSION) {
```

Comment:
Yes, above 'configuration.setVersion()' statement is correct because it is configuring marshalling version and not JNDI client protocol version @chengfang . Also note that JNDI client protocol version can be configured via protocol itself and so it is not guaranteed the version is 3 or above.